src/v/serde/envelope_for_each_field.h

```diff
@@ -212,11 +201,14 @@ constexpr inline auto envelope_to_tuple(T& t) {
     }
 }
 
-template<typename T, typename Fn>
-inline auto envelope_for_each_field(T& t, Fn&& fn) -> std::enable_if_t<
-  !std::is_convertible_v<decltype(fn(std::declval<int&>())), bool>> {
-    static_assert(is_envelope_v<std::decay_t<T>>);
-    if constexpr (inherits_from_envelope_v<std::decay_t<T>>) {
+template<typename Fn>
+concept check_for_more_fn = requires(Fn&& fn, int& f) {
+    { fn(f) } -> std::convertible_to<bool>;
+};
```

Comment:
Is this close enough to std::predicate?

Reply:
yeah, i created a dedicated commit to address your comment. to make the change small enough and more explicit.